### ISSUE TITLE
Magit stage region (selected lines) can lead to git patch corruption

### ISSUE THREAD

**Issue by SteveJobzniak** [magit/magit#3182]
This information comes from a discussion in #3075 (which was about a different line-offset calculation bug, which is what I suspect in this ticket too). I'm making a separate ticket to help you track it.

Back in March, I began using Magit. Initially I loved the ability to stage only a few lines, so that I could stage for example a single changed function within a large file of changes. It worked well for the most part. But then a project collaborator told me that I had broken the repository. And indeed, Github showed the commit as having corrupted lines (some duplicated lines, some removed, etc), around the regions where I had told Magit to stage only a few lines.

Since then, I stopped using the feature (literally within days of becoming a Magit user). I was too sick and had too little energy to report it to you back then.

I have now looked at all Magit changelogs from December of 2016 just since I was curious if it has been fixed, and I saw that nothing relates to this bug.

Here's how the bug works:

- Select a range of lines (within a hunk) and press `s` to stage them. Don't stage the whole hunk. Just a few of its lines.
- Do that a few times for different lines inside the file at various places...
- And then commit.
- The resulting git commit will sometimes be corrupt (magit told Git to stage the wrong lines) and will destroy the target-file once committed, causing stuff like missing lines + duplicated lines...

It happened in Spacemacs with Magit + Evil mode (in March) if that matters.

The broken commit is still out there in the world, but it's hidden somewhere under 1000 commits so I can't find it: https://github.com/mgp25/Instagram-API/commits/master

I am also too sick to provide more information. This is literally all I know about it: I selected a range of lines to stage a small part of a hunk. Magit displayed what looked like a good result. Git itself committed a corrupt change to the file. And it doesn't happen every time the select-lines feature is used.

----

Actually, come to think of it... I wonder if the situation can happen like this: Open Magit window (it will query `git diff`). Open separate window with the same file (within the same Emacs process). Do an edit to that file in the other window (remove or add a line at the top of the file). Go back to the Magit window, select a few lines from the diff (without `gr` refreshing it!). And then stage those lines.

It's not impossible that I accidentally did something like that once. Which would explain why Magit staged the wrong range and corrupted (deleted some lines + duplicated some lines) the file.

**Comment by npostavs** [comment]
> The broken commit is still out there in the world, but it's hidden somewhere under 1000 commits so I can't find it: https://github.com/mgp25/Instagram-API/commits/master

Hmm, you were busy in March, 340 commit mostly yours: https://github.com/mgp25/Instagram-API/compare/master@%7Bmarch%201%7D...master@%7Bmarch%2031%7D. 

mgp25/Instagram-API@eeccd356e84805f8a548123a53cfee5113bb366c looks likely (as the fix to the broken commit) although there's only a single block of duplicated lines.

**Comment by SteveJobzniak** [comment]
@npostavs Holy crap, you found it! Yeah, I was "busy" in March. I joined one of the sloppiest spaghetti projects I've ever seen and went to town refactoring everything. The project went from garbage to winning constant praise now. ;-)

Here's the broken commit (scroll down to line 532 in its list of changes):

https://github.com/mgp25/Instagram-API/commit/53de976e1c45dad28e5e90762f596649dcd34974

Here was the fix (as you found):

https://github.com/mgp25/Instagram-API/commit/eeccd356e84805f8a548123a53cfee5113bb366c

All of that happened by using Magit to select a range around the code to commit, and hitting `s` to stage those lines. I'd been doing that many times (perhaps 20 commits before that used that technique). But then I got a nice message telling me I had broken the repo. Took one look at the commit and went "how on earth did Git do that!?"... ;-)

Sorry that I don't have more details. But I am totally sure that I didn't manually make that mistake of duplicating the body of the function. I never used any chunk-copy/chunk-paste, so I couldn't have made the mistake. And I always reviewed each selection before committing to be sure the code was fine. It really was fine while I was staging... Magit or Git screwed up.

**Comment by tarsius** [comment]
I strongly doubt that Magit (or Git) messed up here^1. This is a very popular feature and I have never heard about something like this happening before. Also, your memory of what you did half a year ago might not be 100% accurate :wink:

It's probably impossible for something like this to happen if all state is up-to-date, i.e. when the information in the file-visiting buffer, the file on disc, and the diff as displayed in Magit are consistent. By default Magit tries hard to keep these things in sync, but it also lets users disable the features that are necessary to ensure this. Have you disabled any of this?

* https://magit.vc/manual/magit/Automatic-Refreshing-of-Magit-Buffers.html
* https://magit.vc/manual/magit/Automatic-Saving-of-File_002dVisiting-Buffers.html
* https://magit.vc/manual/magit/Automatic-Reverting-of-File_002dVisiting-Buffers.html

^1: I think it is impossible that the correct information was feed to Magit and then Git and one of them somehow did disregard that and did something else. It is however possible that outdated information was used and that Magit and/or Git did their best to make sense of it.

Even then though I think staging would fail, while other "apply variants" (apply, revert, discard) might succeed in doing the wrong thing, because they in some cases must be able to apply changes that don't apply exactly.

Staging doesn't need that flexibility. Though I did not study the code enough just now to completely rule out that it does not accidentally use "lax apply". On the other hand, can you rule out that you used only stage, but not one of the other apply variants?

I think that this was caused by a user error. But Magit might have contributed to the issue like so:

* For some of the consistency features above Magit by default asks the user whether s/he wants to do it, giving you the option to do the wrong thing. I don't think that's a good default (because this issue) but I also cannot change it because some vocal users think it is very wrong for Magit to do it in the first place and "at the very least it should ask instead of just doing it".
* If you display the unstaged changes in a diff buffer (as opposed to just looking at them in the status buffer), then it can happen that this information gets out of date, because unlike the status buffer it does not get updated when you invoke a Magit command from another buffer. In the long run I want to fix that, but cannot do that now for performance reasons.

I would like to help you here, regaining your confidence in that Magit will always do the right thing, but without exact steps to reproduce, I don't see how. All in all, I think this was a freak occurrence that is highly unlikely to repeat. I would recommend that you start using this feature again, but carefully look at the diff that is displayed while committing (that's a good idea anyway, as it allows you to spot errors regardless of whether Magit or you made them). 

By the way, I do plan to put "state" under tighter control, though mostly for performance reason. If we can be absolutely sure that some cached state is accurate, then there is no need to re-calculate it. That will a while though.

**Comment by SteveJobzniak** [comment]
Sorry for the late reply. I'm sick.

> This is a very popular feature and I have never heard about something like this happening before.

That's great because I have wished I could use it again, but didn't dare because of that corruption.

> Even then though I think staging would fail, while other "apply variants" (apply, revert, discard) might succeed in doing the wrong thing, because they in some cases must be able to apply changes that don't apply exactly.

Ah, so you're saying that magit makes a diff of the file contents + the changed region and gives the diff to Git, and then Git ensures that it applies exactly as provided by Magit?

If so then I would trust the feature again, because that sounds safe.

You also suggested that I begin the commit and then review the lines a 2nd time. I usually do that too. But I wonder: If Magit/Git was responsible and manages to screw up the commit-staging again, will my file be okay if I just unstage and begin again? It should, but just asking...

> Staging doesn't need that flexibility. Though I did not study the code enough just now to completely rule out that it does not accidentally use "lax apply". On the other hand, can you rule out that you used only stage, but not one of the other apply variants?

Here's the process I used:

- Save file (in Emacs).
- Open Magit (in Emacs).
- Tab to see the diff contents.
- Carefully read the changes (I do this every time I make any commit in any project because ironically enough I hate broken commits ;)).
- Shift-v to select lines.
- Drag a selection with downarrow.
- Press `s` to stage.
- Press `c c` to commit.

That was the process I used every time. I never pressed any other keys.

But I think these scenarios are all possible:

- I may have had some sort of desync between the magit buffer and the actual on-disk file.
- I may actually have hit some "copy region, paste region" command. But I am nearly sure that I didn't do that. I remember the night very well, since it left a scar in my mind never to use the feature again: I looked at the changes (they mainly affected the function signature), selected that region of lines, and staged them.
- Maybe a bug in Spacemacs. It has like 300 packages by default and was under constant development. It has definitely had a fair share of annoying, serious bugs during my time with it. But it had become very stable by now.

So I guess I should just update all Emacs packages and begin using line-by-line staging again! :)

**Comment by tarsius** [comment]
> Ah, so you're saying that magit makes a diff of the file contents + the changed region and gives the diff to Git, and then Git ensures that it applies exactly as provided by Magit?

Yes.

> If Magit/Git was responsible and manages to screw up the commit-staging again, will my file be okay if I just unstage and begin again?

Yes, un-/staging doesn't touch the file(s) at all.

> So I guess I should just update all Emacs packages and begin using line-by-line staging again! :)

Yes!

**Comment by SteveJobzniak** [comment]
@tarsius That's brilliant. The patch-system makes me trust it completely since it means it's not some basic "stage line 11 to 12". It has actual context. I will start using the feature.

And I have updated all my Emacs packages. :-)

> Yes, un-/staging doesn't touch the file(s) at all.

Great to hear... OH YEAH! THAT REMINDS ME OF A DETAIL!

This file wasn't corrupt on-disk. It was only corrupt inside git's repo. And "git status" showed NO changes.

- So I staged a few lines (selected properly).
- Git accepted the patch from Magit.
- I uploaded the commit to the public repo.
- Minutes later I was warned by a project member that I had corrupted the repo.
- I looked at my own file on my disk: `No corruption` !
- I looked at `git status`: No modifications to show (file "up to date"). (Perhaps git had cached the last-modified-time of the file and believed the on-disk file was 100% identical, without it needing to re-parse it for changes).

I literally had to delete my .git folder, re-download the repo, and then do "git status" again to make it detect that the in-repo view of the file was different. Then I staged MY file (the uncorrupt one that I had written), and made the fix-commit.

It seems like `git` was responsible. But I just thought it was magit's fault for "giving git bad data".

Either way, since so many people use the feature, I will just update git + magit to the latest packages and start using it again.

Any final thoughts before we close this ticket? :-)

**Comment by npostavs** [comment]
I tried checking out mgp25/Instagram-API@eeccd35, soft-reset to mgp25/Instagram-API@cb81c37, then reconstructed the next 2 commits by staging all hunks using per-line staging (I just went in order, from top to bottom). I didn't get any corruption.

**Comment by SteveJobzniak** [comment]
@npostavs Wow, excellent initiative to do that test! And great results! Thank you so much for checking that out! Then we can rule out that it'll happen again.

And what I said above in my latest post really was correct. I remember it clearly now. `git push` pushed the corrupt commit online, and `git status` showed no local changes anymore. I had to `rm -rf .git` and `git clone ...` it again, then `git status` to make it detect my current worktree-file. So it really must have been `git` that screwed this up, rather than magit.

What is your `git --version`?  There has been like 15 new versions of Git since I had the bug in March (and I wasn't running the latest Git at that time either since I am on a Mac and only update homebrew packages occasionally).

And each of those new git versions has [huge changelogs](https://github.com/git/git/tree/master/Documentation/RelNotes), so I'll just assume it's fine now. Whatever it was, it must be either ultra-rare or fully fixed, and has nothing to do with Magit either way.

That means we can close this ticket as soon as @tarsius responds about the suggestion to include more lines in the udiff that Magit generates. :-) May be a good idea either way.

**Comment by npostavs** [comment]
Yeah, it seems unlikely that we'll get a reproducible test case from this. But if you do hit this again, don't `rm` your `.git` directory, archive it to look at for later.

```
~/src/Instagram-API$ git --version
git version 2.11.0
```

**Comment by SteveJobzniak** [comment]
Yeah, but from what you guys are saying this is a very popular feature, so it seems unlikely that I'll hit the situation again. I'm just happy that I can finally start line-staging without worrying, like the rest of y'all. ;-)

**Comment by SteveJobzniak** [comment]
@tarsius I don't really understand what you just said, but you know the magit internals so I trust you. ;-) 👍 

Thanks everyone for the discussion.

**Comment by tarsius** [comment]
Lol. Yeah, just saying that we have a diff and want to apply some of that. Taking another diff and applying some of that instead won't make things easier.

**Comment by SteveJobzniak** [comment]
Ah, understood. :-)

**Comment by SteveJobzniak** [comment]
@tarsius Actually, perhaps it's not so hard:

`git diff -U10`

Makes Git generate a diff with 10 lines of context, for example.

With 5 lines I think Magit would be 100% safe against generating bad patches.

**Comment by tarsius** [comment]
Yes of course, but then you also get to see 10 lines of context.

(Use `+` and `-` to quickly change how much context is shown.)

**Comment by SteveJobzniak** [comment]
Hehe okay. I trust you. :-)
